### DISCUSSION TITLE
Support for describing OpenAPI 3.1.0 webhooks

### DISCUSSION THREAD

**Discussion by michaeloliverx** [fastapi/fastapi#8439] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Example taken from https://github.com/OAI/OpenAPI-Specification/blob/main/examples/v3.1/webhook-example.yaml
openapi: 3.1.0
info:
  title: Webhook Example
  version: 1.0.0
# Since OAS 3.1.0 the paths element isn't necessary. Now a valid OpenAPI Document can describe only paths, webhooks, or even only reusable components
webhooks:
  # Each webhook needs a name
  newPet:
    # This is a Path Item Object, the only difference is that the request is initiated by the API provider
    post:
      requestBody:
        description: Information about a new pet in the system
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/Pet"
      responses:
        "200":
          description: Return a 200 status to indicate that the data was received successfully

components:
  schemas:
    Pet:
      required:
        - id
        - name
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
        tag:
          type: string
```

### Description

I would like to be able to describe webhooks in FastAPI as described in [OpenAPI 3.1 Webhooks](https://spec.openapis.org/oas/v3.1.0#oasWebhooks).

## References
- https://spec.openapis.org/oas/v3.1.0#oasWebhooks
- https://github.com/OAI/OpenAPI-Specification/blob/main/examples/v3.1/webhook-example.yaml

### Wanted Solution

I would like an official way to use fastapi idioms to describe these webhooks and potentially add a way to hook into the setup/orchestration of said hooks.

### Wanted Code

```python
from fastapi import FastAPI

app = FastAPI()


# Some way to describe webhooks...
```

### Operating System

Linux, Windows, macOS

### FastAPI Version

0.75.0

### Python Version

3.9.0

**Comment by raphaelauv** [comment]
-> https://github.com/tiangolo/fastapi/pull/3038

**Comment by Kludex** [comment]
I was thinking about this.

How does it fit in FastAPI itself? Like, all the endpoints we build are not webhooks. Unless you use FastAPI applications to build showcase `openapi.json`s... 🤔 

Am I missing something?

**Comment by michaeloliverx** [comment]
> I was thinking about this.
> 
> How does it fit in FastAPI itself? Like, all the endpoints we build are not webhooks. Unless you use FastAPI applications to build showcase `openapi.json`s... 🤔
> 
> Am I missing something?

I was thinking something along the lines of the existing [OpenAPI Callbacks](https://fastapi.tiangolo.com/advanced/openapi-callbacks) support:

> It doesn't need to have any actual code, because your app will never call this code. It's only used to document the external API. So, the function could just have pass.


Ideally I would like a similar way to document OpenAPI Webhooks using pydantic models, docstrings etc and have fastapi generate the appropriate openapi.json for me.

**Comment by michaeloliverx** [comment]
Until Swagger UI is updated to support OpenAPI 3.1 the Swagger docs don't work.. but Redoc does and displays it under an EVENT tag. See example https://redocly.github.io/redoc/#tag/pet/operation/newPet

**Comment by Kludex** [comment]
I see. It makes sense. 👍 

Thanks @michaeloliverx 🙏

**Comment by michaeloliverx** [comment, marked as answer]
With the release of `0.99` this is now possible!
https://github.com/tiangolo/fastapi/releases/tag/0.99.0
https://fastapi.tiangolo.com/advanced/openapi-webhooks/
